### DISCUSSION TITLE
[BUG] callback with type Enum passes None back to the app command

### DISCUSSION THREAD

**Discussion by larsclaussen** [fastapi/typer#1758] [Questions]
### Describe the bug

When adding a callback to a `typer.Option` of type `Enum` the value is not passed back to the app command.

### To Reproduce

Fairly easy to reproduce 

* Create a file `main.py` with:

```Python
import typer
from enum import Enum

app = typer.Typer()


class EndpointChoices(str, Enum):
    localhost = "localhost"
    staging = "staging"
    production = "production"


def models_callback(value: EndpointChoices):
    typer.echo(f"Validating param: {value}, {type(value)}")
    return value


@app.command()
def models(endpoint: EndpointChoices = typer.Option(
            ..., case_sensitive=False, callback=models_callback
        )
):
    """List available models"""
    raise Exception(endpoint)

if __name__ == "__main__":
    app()
```

* Call it with:

```bash
python main.py --endpoint localhost
```

* It outputs:

```
raise Exception(endpoint)
Exception: None

```

* But I expected it to output:

```
localhost, <enum 'EndpointChoices'>
```

### Expected behavior

I would expect the option argument to be present in the scope of the app command.
### Environment

* OS: Ubuntu 20.04
* Python 3.8.5
* Typer Version 0.3.2

**Comment by larsclaussen** [comment]
Just had the idea to return the value of the `EndpointChoices` instance from the `models_callback` like so

```python
def models_callback(value: EndpointChoices):
    typer.echo(f"Validating param: {value}, {type(value)}")
    return value.value
````

Which gives me the result I was expecting 

```python
    raise Exception(type(endpoint))
Exception: <enum 'EndpointChoices'>
```

I'm not sure if this is the expected behavior, though.

**Comment by daddycocoaman** [comment]
Looks like what's happening is that at some point during the callback return, the Enum convertor is being called but it passes in the entire Enum object instead of just the value.  In this case, it's passing in the `EndpointChoices` object instead of `localhost`. 

Here is the Enum being passed in the @app.command and returning the expected value:
![image](https://user-images.githubusercontent.com/21189155/104970509-b971d900-59b9-11eb-8871-205000bcf4f6.png)

Here is the Enum being returned from the callback into the @app.command. Notice the difference in that it's no longer "localhost" being passed, but "EndpointChoices.localhost" as a string, therefore failing the rest of the function and returning `None`. This is why returning `value.value` worked instead...because instead of returning the value of the Enum that @app.command already checked, the callback is returning the entire object.

![image](https://user-images.githubusercontent.com/21189155/104970715-75330880-59ba-11eb-8463-1aa133c1f9c6.png)

The [docs here](https://typer.tiangolo.com/tutorial/options/callback-and-context/#value-function-parameter) indicated this is expected behavior, since your callback should receive the same annotation of the main function! However, I think it's counter-intuitive to not handle that case and return `None`....even as an error it should return the Enum object or raise an error.
